### PR TITLE
gitserver: ReposStats gRPC migration

### PR DESCRIPTION
**Pull request by mucles**
This PR migrates RepoStats in gitserver to use gRPC instead of HTTP when it is enabled. This change is part of an effort to consolidate all internal communication in Sourcegraph to use gRPC.

**Changes**

- [x]  Define ReposStats method in the GitserverService gRPC service in internal/gitserver/v1/gitserver.proto
- [x]  Generate Go code for the gRPC service using sg generate in internal/gitserver/v1/
- [x]  Implement ReposStats client in internal/gitserver/client.go
- [x] implements ReposStats server in cmd/gitserver/server/server_grpc.go
- [x]  Implement ReposStats, the use of a spy to test whether or not the grpc ReposStats client was being called or not changes can be see in internal/gitserver/client.go and internal/gitserver/client_test.go
- [x]  Implement ReposStats, the use of a spy to test whether or not the grpc ReposStats client was being called or not changes can be seen in internal/gitserver/client.go and internal/gitserver/client_test.go

## Test plan
unit test

### CODE REVIEW

File: internal/gitserver/v1/gitserver.proto [hunk 1]
```diff
@@ -210,3 +211,15 @@ message ArchiveRequest {
 message ArchiveResponse {
   bytes data = 1;
 }
+
+message ReposStatsRequest {}
+
+// ReposStats is an aggregation of statistics from a gitserver.
+message ReposStatsResponse {
+  // git_dir_bytes is the amount of bytes stored in .git directories.
+  int64 git_dir_bytes = 1;
```
Comment (ggilmore) [inline internal/gitserver/v1/gitserver.proto:220]: A size can't be negative, right? So, I think it's better to use an unsigned number in this case (`uint64`).

File: internal/gitserver/client_test.go [hunk 2]
```diff
@@ -735,10 +735,85 @@ func TestClient_ReposStats(t *testing.T) {
 	assert.Equal(t, wantStats, *gotStatsMap[gitserverAddr])
 }
 
+func TestClient_ReposStats_GRPC(t *testing.T) {
+	t.Setenv("SG_FEATURE_FLAG_GRPC", "true")
```
Comment (ggilmore) [inline internal/gitserver/client_test.go:739]: If you're setting true here, we should explicitly set false on the other one.

File: internal/gitserver/client_test.go [hunk 3]
```diff
@@ -735,10 +735,85 @@ func TestClient_ReposStats(t *testing.T) {
...
+			return "", nil
+		},
+		GetVCSSyncer: func(ctx context.Context, name api.RepoName) (server.VCSSyncer, error) {
+			return &server.GitRepoSyncer{}, nil
+		},
+	}
+
+	grpcServer := defaults.NewServer(logtest.Scoped(t))
+
+	proto.RegisterGitserverServiceServer(grpcServer, &server.GRPCServer{Server: s})
```
Comment (ggilmore) [inline internal/gitserver/client_test.go:756]: In general, this approach works but the scope of this test is actually much larger than the HTTP version. 

In the HTTP test, we provide a mock client implementation that returns a baked response - we don't actually "spin-up" a HTTP server at all. 

If you wanted to mimic the same approach, we can write a version of the gitserver service client that does the same thing - return a baked response (since all we have to do is fulfill that interface). We're only testing the happy-path in these tests, but mimicking the client directly could be a more-straightfoward way of measuring edge-cases (errors in marshalling, etc.) 

I don't think this approach is bad though - it's testing things more e2e. Just some thoughts if you wanted to evolve the tests here. cc @camdencheek for more thoughts.

**Review by ggilmore — COMMENTED** [review]
Good so far!

File: internal/gitserver/addrs.go [hunk 4]
```diff
@@ -175,6 +184,14 @@ func (a *atomicGitServerConns) ClientForRepo(userAgent string, repo api.RepoName
 	return proto.NewGitserverServiceClient(conn), nil
 }
 
+func (a *atomicGitServerConns) ClientForAddr(userAgent string, addr string) (proto.GitserverServiceClient, error) {
```
[the remark this answers is not in the document]
Comment (mucles) [reply internal/gitserver/addrs.go:187]: Yes, this a great idea; I will work on getting this written out, thank you for the insight @ggilmore  🙏

File: internal/gitserver/client_test.go (hunk 2 above)
Comment (mucles) [reply internal/gitserver/client_test.go:739]: Yes I will explicitly set it to false as well

File: internal/gitserver/client_test.go (hunk 3 above)
Comment (mucles) [reply internal/gitserver/client_test.go:756]: Could we keep this test as the e2e and have another one that mimics the same test as the HTTP with a mock client implementation for gRPC ?

File: cmd/gitserver/server/server_grpc.go [hunk 5]
```diff
@@ -168,3 +170,17 @@ func (gs *GRPCServer) Search(req *proto.SearchRequest, ss proto.GitserverService
...
+func (gs *GRPCServer) ReposStats(ctx context.Context, req *proto.ReposStatsRequest) (*proto.ReposStatsResponse, error) {
+	b, err := gs.Server.readReposStatsFile(filepath.Join(gs.Server.ReposDir, reposStatsName))
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to read %s: %v", reposStatsName, err.Error())
+	}
+
+	var stats *protocol.ReposStats
+	if err := json.Unmarshal(b, &stats); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to unmarshal %s: %v", reposStatsName, err.Error())
+	}
```
Comment (camdencheek) [inline cmd/gitserver/server/server_grpc.go:183]: nit: `err.Error()` is a string value, so we can use the more specific `%s`.

```suggestion
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to read %s: %s", reposStatsName, err.Error())
	}

	var stats *protocol.ReposStats
	if err := json.Unmarshal(b, &stats); err != nil {
		return nil, status.Errorf(codes.Internal, "failed to unmarshal %s: %s", reposStatsName, err.Error())
	}
```

File: internal/gitserver/addrs.go [hunk 6]
```diff
@@ -52,6 +52,7 @@ type TestClientSourceOptions struct {
 }
 
 func NewTestClientSource(addrs []string, options ...func(o *TestClientSourceOptions)) ClientSource {
+	logger := log.Scoped("gitserver", "test client source")
```
Comment (camdencheek) [inline internal/gitserver/addrs.go:55]: nit: would probably be best to use `logtest.Scoped()` here so the logs get forwarded to the `testing.T` object.

File: internal/gitserver/client.go [hunk 7]
```diff
@@ -446,7 +446,12 @@ type Client interface {
 }
 
 func (c *clientImplementor) Addrs() []string {
-	return c.clientSource.Addresses()
+	address := c.clientSource.Addresses()
+	var addrs []string
```
Comment (camdencheek) [inline internal/gitserver/client.go:450]:
```suggestion
	addrs := make([]string, 0, len(address))
```

File: internal/gitserver/client.go [hunk 8]
```diff
@@ -1230,12 +1235,30 @@ func (c *clientImplementor) RepoCloneProgress(ctx context.Context, repos ...api.
...
+			rs := &protocol.ReposStats{}
+			resp, err := client.ReposStats(ctx, &proto.ReposStatsRequest{})
+			if err != nil {
+				allErr = errors.Append(allErr, err)
+			} else {
+				rs.FromProto(resp)
+				stats[addr.Address()] = rs
+			}
+			continue
+		}
```
Comment (camdencheek) [inline internal/gitserver/client.go:1255]: nit: can we move the `IsGRPCEnabled(ctx)` check outside the loop? The control flow is a little difficult to follow with the `continue` skipping all the non-gRPC stuff

File: internal/gitserver/client_test.go [hunk 9]
```diff
@@ -699,6 +736,8 @@ func TestLocalGitCommand(t *testing.T) {
 }
 
 func TestClient_ReposStats(t *testing.T) {
+	t.Setenv("SG_FEATURE_FLAG_GRPC", "false")
```
Comment (camdencheek) [inline internal/gitserver/client_test.go:739]: For new tests, we should prefer using `conf.Mock` to enable/disable gRPC. The environment variable mostly only exists for convenience during dev

File: internal/gitserver/client_test.go [hunk 10]
```diff
@@ -734,13 +773,73 @@ func TestClient_ReposStats(t *testing.T) {
...
+				if err != nil {
+					return nil, err
+				}
+
+				var rs protocol.ReposStats
+				err = json.Unmarshal(encoded, &rs)
+				if err != nil {
+					return nil, err
+				}
+				return rs.ToProto(), nil
```
Comment (camdencheek) [inline internal/gitserver/client_test.go:817]: Since we're mocking out the body of `RepoStats`, I'm not sure there is much value in reading/writing a file. Maybe just `return wantStats.ToProto()` instead to make it more clear what this test actually exercises?

File: internal/gitserver/client.go [hunk 11]
```diff
@@ -1230,12 +1235,30 @@ func (c *clientImplementor) RepoCloneProgress(ctx context.Context, repos ...api.
...
-		stat, err := c.doReposStats(ctx, addr)
+
+	for _, addr := range c.clientSource.Addresses() {
+		if internalgrpc.IsGRPCEnabled(ctx) {
+			client, err := addr.GRPCClient()
+			if err != nil {
+				return nil, err
+			}
+
+			rs := &protocol.ReposStats{}
```
Comment (camdencheek) [inline internal/gitserver/client.go:1246]: nit: this can be defined right before `rs.FromProto()`

**Review by camdencheek — APPROVED** [review]
All comments are nits. This is looking great after resolving Geoffrey's comments!

File: internal/gitserver/client.go (hunk 8 above)
Comment (mucles) [reply internal/gitserver/client.go:1255]: Would you recommend separating the http and grpc flow into there own for loops?

```
if Grpc true

for _, addr := range c.clientSource.Addresses() {
// do grpc stuff
}

else 
for _, addr := range c.clientSource.Addresses() {
// do http stuff
}
```

Comment (camdencheek) [reply internal/gitserver/client.go]: Yeah, I think I would find that a little more clear. Additionally, it means we can get rid of the HTTP side of things just by deleting the branch.
